Create cross codegen extensible peephole phase

fjeremic opened the pull request:
First off, I apologize for the size of this PR. It was a trade off between getting a large scale change done vs. the churn it would have caused if it was done piecemeal. I have split up the logical changes of this work into individual commits, each of which is very consumable from a review point of view.

The changes within this PR are largely cosmetic, as each individual commit shows. Most of it is code moving around the codebase. I'll summarize what was done in chronological bullet point form:

- Removed pre-RA peephole on Z codegen
    - There is already a way to tell whether we are running peepholes before or after RA, so no need to complicate things with an extra codegen phase just to carry out one peephole
- New `OMRPeephole` extensible codegen phase is created which is cross-platform
- All Z peepholes are migrated to the new codegen phase
    - Liberty was taken to clean up code while this was done and I've tried to keep to only simple non-invasive changes
- All Z peepholes have been fully documented
- OpenJ9 peepholes from Z have been migrated to the OpenJ9 extensible peephole and the code no longer lives in OMR
- Fixes for `remove` API to deal with new peephole
- Power, AArch64, and ARM overloads of `remove` have been deprecated in favor of the cross platform implementation
- XPLINK call descriptor compilation failure is now a fatal assert
- Power `OMRPeephole` extensible phase was introduced
- All Power peepholes are migrated to the new codegen phase
- All Power peepholes have been fully documented
- Improvements to tracing and dumping of instructions following peephole passes
- `doPeephole` codegen API removed as it is no longer used
- Added option to disable cross-platform peephole phase

Fixes: #254
Fixes: #5129
Fixes: #5141

fjeremic commented:
Suggesting @0xdaryl for ARM/AArch64 review (this is a very minor piece), @aviansie-ben or @gita-omr  for the Power pieces, and @r30shah for the Z pieces. @klangman FYI as 5cd94d4 was related to a request from you.

fjeremic commented:
@genie-omr build all

fjeremic commented on compiler/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,80 @@
+   TR::Instruction* prevInst = NULL;
+   TR::Instruction* currInst = self()->cg()->getFirstInstruction();
+
+   while (currInst != NULL)
+      {
+      performed |= self()->performOnInstruction(currInst);
+
+      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
+      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
+      if (currInst->getPrev() == prevInst)
+         {
+         prevInst = currInst;
+         currInst = currInst->getNext();
+         }
+      else
+         {
+         currInst = prevInst->getNext();
+         }
+      }
```
How much is "earlier"? Right now if any transformation is performed, i.e. even if some further down instruction is modified, we go back one instruction and restart the peephole there. So for example say we had the following sequence of instructions:

```
X1 
X2 <-- Currently about to call `performOnInstruction` on this one
X3
X4
```

Say `performOnInstruction` returned true, and the instruction sequence following the return looks like:

```
X1
X2
X3
Z4 <-- Got modified to Z4
```

Because `performOnInstruction` returned true, the above code would restart the peephole window at `X1` rather than `X3`, irregardless of what transformation was actually performed. This is because peephole windows only slide down, and peephole should only be modifying instructions following the current instruction being examined, never before!

This should cover the case where new opportunities are unlocked, and it also shields us from `X2` being completely removed from the instruction stream for example.

aviansie-ben replied:
I don't think going back one instruction is sufficient in all cases, though I concede that it may work given our current set of peepholes. Imagine a theoretical situation with the following peepholes:

1. Any time the sequence `C`, `D` shows up, replace it with `E`
2. Any time the sequence `A`, `B`, `E` shows up, replace it with `F`

Now imagine running those peepholes on the following set of instructions:

```
--> A
    B
    C
    D
```

When we first reach `A`, peephole 2 cannot be used, since the instruction after the next one is `C`, not `E`. We then reach `C` and apply peephole 1 to turn the sequence into:

```
    A
--> B
    E
```

However, by only going back one instruction, we don't ever actually end up going back to `A` and thus miss the opportunity to replace the entire sequence with a single instruction `F`.

fjeremic replied:
Right for sure there will be missed opportunities. Same issue happens with the simplifier. The way we overcome that in the optimizer is by running the opt multiple times. In general there will be no way to know how far back we have to go to ensure we catch the opportunity you describe because that would involve peephole 1. being aware of peephole 2., and in general all other peepholes as well. This doesn't seem feasible.

We can run the peephole iteratively until the `perform()` function returns false, however the complexity of such algorithm is at least O(n^2) where n is the number of instructions. IMO the peephole pass should be quick and simple reductions. We would probably catch a large majority of cases by running peepholes twice if need be.

aviansie-ben replied:
Fair enough, I'm fine with whatever we come up with here (even if it's not perfect) as long as the issue has been considered and what we have handles the majority of the cases we want to optimize. You'd need to either keep running peepholes until nothing happens or back up by the maximum window size of a peephole to fully solve the problem, which may not be an acceptable cost. This was just a thought I had while reading the code after it was moved.

fjeremic commented on compiler/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,80 @@
+
+      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
+      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
+      if (currInst->getPrev() == prevInst)
+         {
+         prevInst = currInst;
+         currInst = currInst->getNext();
+         }
+      else
+         {
+         currInst = prevInst->getNext();
+         }
```
> Do we need to document a set of assumptions that peepholes must obey for this code to work?

I think we should make our life simple and just document the restriction. Peepholes should _only_ modify instructions following (or at) the current instruction being examined, never before. This is currently true for all peepholes that exist in the codebase today.

fjeremic replied:
Fixed in b73252e.

aviansie-ben replied:
I'm not sure this is actually true. `recordFormPeephole` in particular will walk backwards after finding a `cmpdi cr0, rX, 0` instruction to try and find the instruction that wrote to rX and might change that instruction to make it record-form. In fact, for `rlwinm` instructions on older versions of Power, it might remove it entirely and replace it with an `andi.` or `andis.` instruction.

fjeremic replied:
You're right. The piece of code in question is in [1], and there is another example just a few lines before. I've searched for all references to `->getPrev` in the Power peepholes and this is the only place where we could potentially remove the previous instruction as well as the current instruction. In such a case the code being reviewed here is problematic as the previous instruction will have been unlinked from the instruction stream.

How to fix this?

1. Expose `prevInst` and `currInst` as fields rather than local variables in the above algorithm, which extensions can modify as they see fit
    - In the Power case we would modify these pointers when we are about to perform the record form peephole so it knows where to restart
2. We augment all peepholes to return the restart point
3. We backpedal on f4d073637e6e751673c0cbd32f98c7dd66cd8b2c and somehow modify this algorithm to handle the pointers not being set to `NULL`
4. ...?

I'm in favor of 1. to handle these cases. I don't know if there are other alternatives that people can think of. 2. seems quite invasive as every peephole would suffer just so we can accommodate this one exceptional case. 3. seems opposite of where we want to be. Why should dangling instructions exist? @aviansie-ben thoughts on the options or alternatives?

[1] https://github.com/eclipse/omr/pull/5235/files#diff-bdc496687bbdf4ac8bd615034145e891R201-R207

fjeremic replied:
Thinking about this some more, option 4. is possible if we do something like:

```
   while (currInst != NULL)
      {
      performed |= self()->performOnInstruction(currInst);

      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
      if (currInst->getPrev()->getNext() == currInst)
         {
         prevInst = currInst;
         currInst = currInst->getNext();
         }
      else
         {
         currInst = currInst->getPrev()->getNext();
         }
      }
```

which may be more appealing. This works even in the case we remove both `currInst` and `prevInst`, because the removal has to happen one instruction at a time. @aviansie-ben do you think we should prefer option 4. in light of this?

aviansie-ben replied:
I'd be alright with exposing `prevInst` as a field to get this PR moved along. In the long term, I'd prefer that peepholes specifically return their restart points (possibly by modifying the `cursor` field?) since having two instruction cursors as fields seems like it could cause confusion IMO. That being said, I don't think this is an issue worth delaying this PR over, since it would involve more invasive changes to the peepholes themselves.

aviansie-ben replied:
Well, it looks like we once again ended up posting comments at nearly exactly the same time... I think the solution you've come up with regarding option 3 might work given the current set of peepholes, but it doesn't work in the general case.

Imagine that we start with a sequence of instructions `A, B, C, D, E` and start a peephole at `C`, so `prevInst = B` and `currInst = C`. We start by removing `C`, leaving `C->getPrev() = B`. Next, we remove `B`, leaving `B->getNext() = D`. From there, we can end up in a problematic state by removing `D`. In this case `C->getPrev()->getNext() = D`, so we end up restarting peepholes starting from the removed instruction `D`. Granted, this is a pretty contrived series of events, but it could still theoretically occur.

fjeremic replied:
> Granted, this is a pretty contrived series of events, but it could still theoretically occur.

I agree. Although very unlikely it is still possible and that is deterrent enough. I'm going to go ahead with implementing 1. in a separate commit so it's easy to review. In the future I also agree we need a proper way for peepholes to indicate a restart point. However we are not there just yet.

fjeremic replied:
@aviansie-ben I've gone ahead and implemented the fix in 79d2fe3. I'm not happy with the solution but it seems to be the least invasive one which doesn't interfere with other peepholes and allows us to proceed with the work here.

I've summarized the problem in the commit message for future reference. If you are ok with this fix I will go ahead and open up an issue and write down some thoughts so we can remove this hack and think about peephole traversals, peephole windows, restart points, etc. Let me know!

fjeremic commented on compiler/codegen/OMRPeephole.hpp:
```diff
@@ -0,0 +1,71 @@
+
+   private:
+
+   TR::Compilation* _comp;
+   TR::CodeGenerator* _cg;
```
You're right, technically we can just grab the compilation from codegen. I was just modeling how other extensible classes looked since some appear to have both. I'd definitely want a second opinion on this (@0xdaryl?), but for the record I'm happy to make the change if we all agree. We should likely make a similar change in other extensible classes in the codegen where this is prevalent. I can open an issue for that if we agree this is the direction we want to take.

fjeremic replied:
@0xdaryl thoughts on the above?

fjeremic replied:
Do you think we should do the work now (just in the Peephole changes) or have a blanked issue to cover all other classes and do them all at once at a later point?

0xdaryl replied:
Not urgent in my mind.  This can be cleaned up later.  I only commented because my opinion was asked. :-)

fjeremic commented on compiler/p/codegen/OMRPeephole.hpp:
```diff
@@ -0,0 +1,195 @@
+   private:
+
+   /// The instruction cursor currently being processed by the peephole optimization
+   TR::Instruction* cursor;
```
I initially had it like that but changed it in 28794988df69c713df6052ed294bad4baee1337f. If you look at that commit you will see there is quite a bit of repetition happening, especially in the documentation. Since all peephole methods are private and not publicly exposed I thought let's avoid the repetition and just cache the value as a private field and simplify all the APIs and documentation.

fjeremic commented on compiler/codegen/OMRCodeGenPhase.cpp:
```diff
@@ -316,16 +317,21 @@ OMR::CodeGenPhase::performBinaryEncodingPhase(TR::CodeGenerator * cg, TR::CodeGe
 
-   TR::LexicalMemProfiler mp(phase->getName(), comp->phaseMemProfiler());
-   LexicalTimer pt(phase->getName(), comp->phaseTimer());
+   if (comp->getOption(TR_DisablePeephole))
```
Yikes! This is a total mistake. Thanks for catching this. For the record I have tested the changes before adding this commit and everything was working stellar, so I don't expect problems once this is fixed.

fjeremic replied:
Fixed in [force-push](https://github.com/eclipse/omr/compare/0a679a6766d2c7c1df96ade29708a00c92ef74bd..b73252e8b2a6a8d4985995ad719b1666197bde9e#diff-e06618ef688bc0368fe19b670a58e71fR322).

fjeremic commented:
@genie-omr build all

0xdaryl commented:
Does this have any downstream dependencies?  Just confirming it can merge on its own.

fjeremic commented:
> Does this have any downstream dependencies? Just confirming it can merge on its own.

Yes it does. See eclipse/openj9#9672.

fjeremic commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,1866 @@
+         windowSize<maxWindowSize)
+      {
+      // does instruction load or store? otherwise just ignore and move to next instruction
+      if (current->isLoad() || current->isStore())
```
Load and test AGI reduction is handled via `tryToReduceAGI`. For the transformation you suggest we currently don't have such a peephole. I would qualify this type of work outside the scope of this PR which does not implement new functionality. If you believe such cases are prevalent we should open up a separate issue which describes the peephole and tracks implementing it.

r30shah replied:
I agree. This is not part of this PR, just a question I had when I go through the changes, wondering if we has some reason behind limiting this.

fjeremic commented on compiler/z/codegen/OMRPeephole.hpp:
```diff
@@ -0,0 +1,493 @@
+
+   /** \brief
+    *     Tries to reduce AGIs (Address Generation Interlock) which occur when an instruction requires a register
+    *     its operand address calculation but the register is unavailable because it is written to by a preceeding
```
That's right. Fix in 81de07d

fjeremic commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,1866 @@
+      reachedBranch = reachedBranch || (current->isBranchOp() && !(current->isExceptBranchOp()));
+
+      // does instruction load or store? otherwise just ignore and move to next instruction
+      if (current->isLoad() || current->isStore())
```
See above for discussion on this topic.

fjeremic commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,1866 @@
+            storeInst->getMemoryReference()->getIndexRegister()->decTotalUseCount();
+            }
+
+         TR::Instruction * newInst = generateSS1Instruction(self()->cg(), TR::InstOpCode::MVC, self()->comp()->getStartTree()->getNode(), size - 1, storeInst->getMemoryReference(), loadInst->getMemoryReference(), cursor->getPrev());
```
This is a mistake in the original implementations. Fixed in 81de07d

fjeremic commented:
@gita-omr @aviansie-ben @r30shah let me know if there are other points of discussion on the changes here. I've completed addressing all review comments. If the review comments you've raised already have been addressed to your satisfaction please go ahead and resolve the review as complete so we can see a clear list of what is left.

I'll be watching any activity here and implementing changes as reviews come in. Thanks!

gita-omr commented:
@fjeremic I have not started my review yet. I will try to do it today or tomorrow.

gita-omr commented:
@fjeremic could you please elaborate on "Fixes for remove API to deal with new peephole"? How does the new peephole deal with removing instructions? Can the change to the old remove() function (setting prev and next to NULL) affect places in the codegens other than peephole that might be using that function as well?

gita-omr commented on compiler/p/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1538,88 +641,6 @@ bool OMR::Power::CodeGenerator::checkAndFetchRequestor(TR::Instruction *instr, T
-
-   while (instructionCursor)
-      {
-      self()->setCurrentBlockIndex(instructionCursor->getBlockIndex());
```
Do we do the same in the new peephole pass?

fjeremic replied:
No, because there is no peephole which consumes this. In fact these APIs should be deprecated as they are currently unused. I've opened #5293 to track this.

aviansie-ben commented on compiler/p/codegen/PPCBinaryEncoding.cpp:
```diff
@@ -902,12 +902,6 @@ uint8_t *TR::PPCAlignmentNopInstruction::generateBinaryEncoding()
 
    cg()->addAccumulatedInstructionLengthError(getEstimatedBinaryLength() - currentMisalign);
 
-   // When the trace log prints the list of instructions after binary encoding, we don't want this
```
This `remove()` call was added for cosmetic reasons. Specifically, as the comment says, to prevent this alignment nop from showing up in tracefile output after it has expanded. While this is nice, it's not really necessary for the code to function correctly (and in fact, I've later realized that it can end up being more confusing than helpful sometimes).

However, one of the assumptions made by this code, namely that `remove()` leaves the `next` field alone so that binary encoding can continue, has now been violated. As a result, this no longer works correctly and can't be left as-is.

fjeremic commented on compiler/codegen/OMRInstruction.cpp:
```diff
@@ -216,25 +216,18 @@ Instruction::move(TR::Instruction *newLocation)
-      {
-      cg->setAppendInstruction(prev);
-      }
+   self()->setPrev(NULL);
```
See the explantion in the commit message of f4d0736 which is the commit that made the change.

fjeremic commented:
> @fjeremic could you please elaborate on "Fixes for remove API to deal with new peephole"? How does the new peephole deal with removing instructions?

In the forward peephole pass, we maintain the previous instruction that we ran the peephole optimization on. If the `performOnInstruction` API returns `true`, it indicates that a peephole transformation was performed. This could mean several things:

- An instruction is modified
- An instruction was moved or removed
- An instruction was added

If _any_ operation was performed we restart the forward peephole pass on the previous instruction, which is always valid. The only backwards peephole pass we have is the Power record form peephole. Because we currently don't have a proper backwards peephole pass we had to do special care to ensure the restart point is valid. See [1] review discussion with @aviansie-ben on this topic and how it was fixed in 79d2fe3. See the commit description for more details.

[1] https://github.com/eclipse/omr/pull/5235#discussion_r430501548

> Can the change to the old remove() function (setting prev and next to NULL) affect places in the codegens other than peephole that might be using that function as well?

I've examined all the places that called `remove()` and determined that setting the prev. and next pointers to `NULL` will not affect those locations. For the one or two places that it will affect, I've changed the code accordingly to handle how the new semantics of the `remove` API. Those changes were made in f4d0736.

gita-omr commented:
I slightly worry of people making mistakes by not returning correct value from performOnInstruction as well as updating previous instruction on the backward path...  Maybe we should document it somewhere in the class?

fjeremic commented:
> Maybe we should document it somewhere in the class?

Yeah it is documented as part of the API documentation:
https://github.com/eclipse/omr/pull/5235/files#diff-f85ab2a045cdbae4c02333f3dd38648dR71-R80

I don't think we can guard against programmer errors not following API spec. Such issues should be caught in reviews though.

fjeremic commented:
Thanks for the review @gita-omr! Much appreciated. @0xdaryl I think we have approval from all suggested reviewers. I'll go ahead and wait for your review as well upon which I'll fix the merge conflict as well.

fjeremic commented:
@0xdaryl gentle remind for this one. There is also a question for you in the above review thread regarding the helper functions.

knn-k commented:
There are merge conflicts in compiler/control/OMROptions.hpp and compiler/p/codegen/OMRCodeGenerator.cpp.

fjeremic commented:
Waiting on @0xdaryl to review still. I'll resolve the conflict momentarily. This PR touches a lot of code so we are going to keep running into these issues.

fjeremic commented:
Rebase completed. Launching builds again to verify. The changes were quite simple, and the conflicting PR was #5197.

@genie-omr build all

fjeremic commented:
Rebase completed. Launching builds again to verify. The changes were quite simple, and the conflicting PR was #5197.

@genie-omr build all

fjeremic commented:
Rebase completed. Launching builds again to verify. The changes were quite simple, and the conflicting PR was #5197.

@genie-omr build all

fjeremic commented:
@0xdaryl a friendly reminder for action on this one.

aviansie-ben commented:
@0xdaryl Because of this PR being in limbo, I'm not currently able to fix #5418 without causing major merge conflicts. That issue is currently blocking completion of pTOC replacement for POWER10 (#5367), so I'd like to see this merged ASAP so that POWER10 work can proceed.

fjeremic commented:
Rebased on the latest master branch. Rebae was clean. Re-launching builds:

@genie-omr build all

dchopra001 commented:
Processor detection is currently not working on JITServer for Z. This PR will need to be merged before those changes can be made. (For now I have a tactical fix that I'll work with).

fjeremic commented:
@0xdaryl sorry to bug you again. We have @aviansie-ben and @dchopra001 blocked by this work. Let me know if there is anything else to address on my front to get this merged.

0xdaryl commented:
Sorry for the delay.  I'll have some time later this afternoon/evening to review.  Been backed up with other things unfortunately.

0xdaryl commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -448,6 +453,92 @@ OMR::Z::Peephole::attemptToReduceAGI(TR::Instruction* cursor)
+   if (isICMOpportunity && performTransformation(comp(), "\nO^O S390 PEEPHOLE: Reducing L [%p] being reduced to ICM.\n", cursor))
+      {
+      // Prevent reuse of memory reference
+      TR::MemoryReference* memcp = generateS390MemoryReference(*load->getMemoryReference(), 0, cg());
```
Use `mem` instead of `load->getMemoryReference()` ?

fjeremic replied:
Fixed in ece2f17

0xdaryl commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -544,6 +549,40 @@ OMR::Z::Peephole::attemptToReduceLToICM(TR::Instruction* cursor)
+      TR::Register *llcTgtReg = ((TR::S390RRInstruction *) cursor)->getRegisterOperand(1);
+
+      TR::Register *curSrcReg = ((TR::S390RRInstruction *) current)->getRegisterOperand(2);
+      TR::Register *curTgtReg = ((TR::S390RRInstruction *) current)->getRegisterOperand(1);
```
Just confirming this is correct as both `curTgtReg` and `llcTgtReg` are operand 1.

0xdaryl replied:
Actually, if that is true then the condition below doesn't need the `llcTgtReg == curTgtReg` check as it will always be true.

fjeremic replied:
Yes those are correct, however the values are not the same. One extracts it from `current` one extracts from `cursor`, and `current = cursor->getNext()`. Poor variable names IMO. I just copied it from existing code and did not make any changes. I've changed the variable name of `current` to `nextInst` so it is more clear. Fixed in 92617ef

0xdaryl replied:
Oh, duh.  Yeah, I missed that.  Sorry.

fjeremic commented on compiler/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,80 @@
+         }
+      else
+         {
+         currInst = prevInst->getNext();
```
I think it is theoretically possible if some peephole removes the first instruction in the instruction stream. I'm not sure how to guard against this, other than to start peepholes on the second instruction in the stream. Any better alternatives?

0xdaryl replied:
Sure, that sounds fine with me.

fjeremic commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -448,6 +453,92 @@ OMR::Z::Peephole::attemptToReduceAGI(TR::Instruction* cursor)
+      isICMOpportunity = true;
+      }
+
+   if (isICMOpportunity && performTransformation(comp(), "\nO^O S390 PEEPHOLE: Reducing L [%p] being reduced to ICM.\n", cursor))
```
Looking at the original code [1] which looks identical, yes this variable is not needed. Fixed in 4b04564

[1] https://github.com/eclipse/omr/blob/deecc7a4f0b3d0d05b47d5b7db35fce3d8f50d0a/compiler/z/codegen/S390Peephole.cpp#L517-L543

fjeremic commented on compiler/z/codegen/OMRPeephole.cpp:
```diff
@@ -0,0 +1,1866 @@
+      TR::S390RXInstruction* storeInst = static_cast<TR::S390RXInstruction*>(cursor->getNext());
+
+      // Cannot change to MVC if symbol references are unresolved, MVC doesn't follow same format as load and store so patching won't work
+      if (loadInst->getMemoryReference()->getSymbolReference() && loadInst->getMemoryReference()->getSymbolReference()->isUnresolved() ||
```
Fixed in ccf148e

fjeremic commented:
Rebase was fairly clean with only one commit needing to get fixed up.

@genie-omr build all

0xdaryl commented:
@genie-omr build zos

fjeremic commented:
@genie-omr build zos
